src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1594,49 +1806,48 @@ final String nextWorkerThreadName() {
     }
 
     /**
-     * Finishes initializing and records owned queue.
+     * Finishes initializing and records internal queue.
      *
      * @param w caller's WorkQueue
      */
     final void registerWorker(WorkQueue w) {
-        ThreadLocalRandom.localInit();
-        int seed = ThreadLocalRandom.getProbe();
-        ReentrantLock lock = registrationLock;
-        int cfg = config & FIFO;
-        if (w != null && lock != null) {
+        if (w != null) {
             w.array = new ForkJoinTask<?>[INITIAL_QUEUE_CAPACITY];
-            cfg |= w.config | SRC;
-            w.stackPred = seed;
-            int id = (seed << 1) | 1;                   // initial index guess
-            lock.lock();
+            ThreadLocalRandom.localInit();
+            int seed = w.stackPred = ThreadLocalRandom.getProbe();
+            int id = ((seed << 1) | 1) & SMASK; // base of linear-probe-like scan
+            int stop = lockRunState() & STOP;
             try {
-                WorkQueue[] qs; int n;                  // find queue index
+                WorkQueue[] qs; int n;
                 if ((qs = queues) != null && (n = qs.length) > 0) {
-                    int k = n, m = n - 1;
-                    for (; qs[id &= m] != null && k > 0; id -= 2, k -= 2);
-                    if (k == 0)
-                        id = n | 1;                     // resize below
-                    w.phase = w.config = id | cfg;      // now publishable
-
+                    for (int k = n, m = n - 1;  ; id += 2) {
```

Comment:
@DougLea Might make sense to create a constant each for worker-queue stride and submission-queue stride so it becomes easier to reason about the `2` here.

Reply:
There seem to be too many variants of queues array traversal loops to do this uniformly, but I did add some explanation:

     * alternatives. Other loops over the queues array vary in origin
     * and stride depending on whether they cover only submission
     * (even) or worker (odd) queues or both, and whether they require
     * randomness (in which case cyclically exhaustive strides may be
     * used).